api/v1alpha1/executionhook_types.go

```diff
@@ -0,0 +1,189 @@
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=executionhook,shortName=eh,scope=Namespaced,categories=executionhook
+// +kubebuilder:subresource:status
+
+// ExecutionHook is the Schema for the executionhook API
+type ExecutionHook struct {
+	metav1.TypeMeta `json:",inline"`
+	// Metadata associated with persisted resources
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// ExecutionHookSpec defines the desired state of ExecutionHook
+	Spec ExecutionHookSpec `json:"spec,omitempty"`
+	// ExecutionHookStatus defines the observed state of ExecutionHook
+	// +optional
+	Status *ExecutionHookStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ExecutionHookList contains a list of ExecutionHook
+type ExecutionHookList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ExecutionHook `json:"items"`
+}
+
+// ExecutionHookSpec defines the desired state of ExecutionHook
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Application Snapshot Controller.
+type ExecutionHookSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// PodSelection defines how to select pods and containers to run
+	// the executionhook. If multiple pod/containers are selected, the action will executed on them
+	// asynchronously. If execution ordering is required, caller has to implement the logic and create
+	// different hooks in order.
+	// This field is required.
+	PodSelection PodSelection `json:"podSelection"`
+
+	// Name of the HookAction. This is required.
+	ActionName string `json:"actionName"`
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where exactly one of them must be defined so that the hook controller knows where to
+// run the hook.
+// TODO: Add validation to ensure exactly one of them are defined.
+type PodSelection struct {
+	// PodContainerNamesList lists the pods/containers on which the ExecutionHook
+	// should be executed. If not specified, the ExecutionHook controller will find
+	// all pods and containers based on PodContainerSelector.
+	// If both PodContainerNamesList and PodContainerSelector are not
+	// specified, the ExecutionHook cannot be executed and it will fail.
+	// +optional
+	PodContainerNamesList []PodContainerNames `json:"podContainerNamesList,omitempty"`
+
+	// PodContainerSelector is for hook controller to find pods and containers
+	// based on the pod label selector and container names
+	// If PodContainerNamesList is specified, this field will not be used.
+	// +optional
+	PodContainerSelector *PodContainerSelector `json:"podContainerSelector,omitempty"`
+}
+
+// PodContainerNames lists the containers the ExecutionHook should be executed on in a Pod.
+type PodContainerNames struct {
+	// This field is required
+	PodName string `json:"podName"`
```

Comment:
This doesn't make a lot of sense to me, if your selector identifies e.g. all pods in a replica set, this still makes you specify each one by name (since the name has a random suffix).

Reply:
Only one of "PodContainerNames" and "PodContainerSelector" will be specified but not both.  So if the selector identifies all pods in the PodContainerSelector, this PodContainerNames won't be used.  We need to add validation to make sure only one is specified.